src/Compilers/Core/Portable/MetadataReader/PEModule.cs

```diff
@@ -1312,37 +1314,195 @@ private ArrayBuilder<string> ExtractStringValuesFromAttributes(List<AttributeInf
                     {
-                        return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message, isError: false);
+                        // check prolog
+                        if (reader.ReadInt16() == 1)
+                        {
+                            var (diagnosticId, urlFormat) = CrackObsoleteProperties(ref reader);
+                            return new ObsoleteAttributeData(ObsoleteAttributeKind.Obsolete, message: null, isError: false, diagnosticId, urlFormat);
+                        }
                     }
+                }
+            }
+            catch (BadImageFormatException)
+            { }
 
-                    return null;
+            return null;
+        }
 
-                case 2:
-                    // ObsoleteAttribute(string, bool)
-                    return TryExtractValueFromAttribute(attributeInfo.Handle, out var obsoleteData, s_attributeObsoleteDataExtractor) ?
-                        obsoleteData :
-                        null;
+        /// <summary>
+        /// Decodes attribute argument type from attribute blob (called FieldOrPropType in the spec).
+        /// </summary>
+        /// <exception cref="UnsupportedSignatureContent">If the encoded argument type is invalid.</exception>
+        /// <exception cref="BadImageFormatException">An exception from metadata reader.</exception>
+        private static (SerializationTypeCode typeCode, SerializationTypeCode elementTypeCode) DecodeCustomAttributeFieldOrPropTypeOrThrow(ref BlobReader argReader)
+        {
+            var typeCode = argReader.ReadSerializationTypeCode();
+
+            // Spec:
+            // The FieldOrPropType (typeCode) shall be exactly one of: ELEMENT_TYPE_BOOLEAN,
+            // ELEMENT_TYPE_CHAR, ELEMENT_TYPE_I1, ELEMENT_TYPE_U1, ELEMENT_TYPE_I2,
+            // ELEMENT_TYPE_U2, ELEMENT_TYPE_I4, ELEMENT_TYPE_U4, ELEMENT_TYPE_I8,
+            // ELEMENT_TYPE_U8, ELEMENT_TYPE_R4, ELEMENT_TYPE_R8, ELEMENT_TYPE_STRING.
+            // 
+            // A single-dimensional, zero-based array is specified as a single byte 0x1D followed
+            // by the FieldOrPropType of the element type. (See §II.23.1.16) An enum is
+            // specified as a single byte 0x55 followed by a SerString.
+            // 
+
+            // todo: this could be simplified for the "skip" case.
+            if (typeCode == SerializationTypeCode.SZArray)
+            {
+                var elementTypeCode = argReader.ReadSerializationTypeCode();
+                if (elementTypeCode == SerializationTypeCode.SZArray)
+                {
+                    // nested array not allowed:
+                    throw new UnsupportedSignatureContent();
+                }
+
+                return (typeCode, elementTypeCode);
+            }
+
+            switch (typeCode)
+            {
+                case SerializationTypeCode.Enum:
+                    if (!CrackStringInAttributeValue(out _, ref argReader))
```

Comment:
>if (!CrackStringInAttributeValue(out _, ref argReader)) [](start = 20, length = 55)

It doesn't look like we can discard the type name because without it we wouldn't know the underlying integer type to skip the value. Unfortunately, this component simply doesn't have enough context information to be able to resolve the string to the enum type. Therefore, I am getting convinced that cracking named args in attributes is rather complicated activity and trying to optimize it is probably not worth the additional complexity. Why don't we indeed try to leverage MetadataDecoder (as you suggested offline)? Let's try changing TryGetDeprecatedOrExperimentalOrObsoleteAttribute to take an instance of MetadataDecoder and leverage helpers it has to crack individual named args.